File: mixin-padang-garut/src/main/python/garut/functions/dataset/replace_value.py
#
# Copyright (c) 2020-2023 Datamixin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
from typing import Dict
import numpy as np

import pandas as pd
from garut.parameter import Parameter
from garut.contexts.context import Context
from garut.instruction import InstructionFactory
from garut.functions.function import FunctionRegistry
from garut.functions.dataset.dataset_function import DatasetFunction


class ReplaceValue(DatasetFunction):

    FUNCTION_NAME = "ReplaceValue"

    COLUMN = "column"
    TARGET = "target"
    REPLACEMENT = "replacement"
    REGEX = "regex"

    def execute(self, context: Context, options: Dict[str, any]) -> any:
        dataset: pd.DataFrame = self.getDataset(options)
        column = options.get(ReplaceValue.COLUMN, None)
        target = options.get(ReplaceValue.TARGET, None)
        replacement = options.get(ReplaceValue.REPLACEMENT, None)
        regex = options.get(ReplaceValue.REGEX, False)
        if dataset[column].dtype.type == str or dataset[column].dtype.type == np.string_:
            result = dataset.copy()
            result[column] = result[column].str.replace(target, replacement, regex=regex)
            return result
        else:
            return dataset.replace({column: target}, {column: replacement}, regex=regex)


parameters = [
    DatasetFunction.DATASET,
    ReplaceValue.COLUMN,
    ReplaceValue.TARGET,
    ReplaceValue.REPLACEMENT,
    Parameter(ReplaceValue.REGEX, bool, False),
]
registry: FunctionRegistry = FunctionRegistry.getInstance()
registry.register(ReplaceValue.FUNCTION_NAME, ReplaceValue, parameters)

factory: InstructionFactory = InstructionFactory.getInstance()
factory.registerInstruction(ReplaceValue.FUNCTION_NAME)

File: mixin-padang-webface/src/main/js/padang/functions/dataset/ReplaceValue.ts
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. */
import FunctionPlan from "padang/plan/FunctionPlan";
import ParameterPlan from "padang/plan/ParameterPlan";
import ParameterPlanUtils from "padang/plan/ParameterPlanUtils";
import FunctionPlanRegistry from "padang/plan/FunctionPlanRegistry";

import ColumnFunction from "padang/functions/dataset/ColumnFunction";
import DatasetFunction from "padang/functions/dataset/DatasetFunction";

export default class ReplaceValue extends DatasetFunction {

    public static FUNCTION_NAME = "ReplaceValue";

    constructor(
        public dataset: string,
        public column: string,
        public target: string,
        public replacement: string,
        public regex: boolean) {
        super(ReplaceValue.FUNCTION_NAME, dataset);
    }

    public static COLUMN_PLAN: ParameterPlan = ParameterPlanUtils.createTextPlan(
        "column",
        "Column",
        "Column where value will be replaced",
        "",
        ColumnFunction.getDefaultColumnKeysAssignable()
    );
    public static TARGET_PLAN: ParameterPlan = ParameterPlanUtils.createTextPlan(
        "target",
        "Target",
        "Target value to be replaced"
    );
    public static REPLACEMENT_PLAN: ParameterPlan = ParameterPlanUtils.createTextPlan(
        "replacement",
        "Replacement",
        "New replacement value"
    );
    public static REGEX_PLAN: ParameterPlan = ParameterPlanUtils.createLogicalPlan(
        "regex",
        "Regex",
        "Use regex"
    );

    public static getPlan(): FunctionPlan {
        let plan = new FunctionPlan(
            ReplaceValue.FUNCTION_NAME,
            "Replace Value",
            "mdi-file-replace-outline",
            "Replace column target value into a new replacement value"
        );
        let parameters = plan.getParameterList();
        parameters.add(DatasetFunction.DATASET_PLAN);
        parameters.add(ReplaceValue.COLUMN_PLAN);
        parameters.add(ReplaceValue.TARGET_PLAN);
        parameters.add(ReplaceValue.REPLACEMENT_PLAN);
        parameters.add(ReplaceValue.REGEX_PLAN);
        return plan;
    }

}

let registry = FunctionPlanRegistry.getInstance();
registry.registerPlan(ReplaceValue.getPlan());